### PR TITLE
Exponential overflow header buffer

### PR DESCRIPTION
**Pull request by kylebarron**
### Changes 

- Updates `AsyncFileReader` to use `&mut self` instead of `&self` on the trait methods. This is necessary so that the buffer can be **mutated** when fetching the overflow, while also adding it to the cached buffers
- Use `Box` instead of `Arc` in reader code, now that we need mutability of the reader.
- Updates the tokio integration to implement `AsyncFileReader` directly on anything that implements `AsyncRead` and `AsyncSeek`. We don't need a wrapper object now that we use `&mut self` ourselves.
- Add test to verify overflow handling

This PR is not made on top of #82, but I think some mix of this and #82 makes sense.

### CODE REVIEW

File: src/reader.rs [hunk 1]
```diff
@@ -227,54 +207,121 @@ impl ReqwestReader {
...
+    async fn expand_prefetch(&mut self, range: Range<u64>) -> AsyncTiffResult<Bytes> {
+        let existing_buffer_length = self.buffer_length() as u64;
+        let additional_fetch =
+            (existing_buffer_length as f64).powf(self.overflow_fetch_exponent) as u64;
+
+        // Make sure that we fetch at least the entire desired range
+        let new_range =
+            existing_buffer_length..range.end.max(existing_buffer_length + additional_fetch);
+        let buffer = self.reader.get_metadata_bytes(new_range).await?;
+        self.buffers.push(buffer);
```
Comment (kylebarron) [inline src/reader.rs:259]: This is where we request the overflow and add it to the existing buffering (`self.buffers.push(buffer)`)

File: src/reader.rs [hunk 2]
```diff
@@ -227,54 +207,121 @@ impl ReqwestReader {
...
+    fn buffer_length(&self) -> usize {
+        self.buffers.iter().fold(0, |acc, x| acc + x.len())
+    }
+
+    /// Access the buffer range out of the cached buffers
+    ///
+    /// ## Panics
+    ///
+    /// If the range does not fall completely within the pre-cached buffers.
+    fn buffer_slice(&self, range: Range<u64>) -> Bytes {
```
Comment (kylebarron) [inline src/reader.rs:275]: This reads from the existing buffers, handling a cross-buffer merge in case the input request overlaps multiple of `self.buffers`

**Comment by kylebarron** [comment]
Closing in favor of #140
